modules/ml/include/opencv2/ml.hpp

```diff
@@ -1838,6 +1867,166 @@ CV_EXPORTS void randMVNormal( InputArray mean, InputArray cov, int nsamples, Out
+*                                   Simulated annealing solver                             *
+\****************************************************************************************/
+
+/** @brief The class implements simulated annealing for optimization.
+@cite Kirkpatrick83 for details
+*/
+class CV_EXPORTS SimulatedAnnealingSolver : public Algorithm
+{
+public:
+    SimulatedAnnealingSolver() { init(); };
+    /** Give energy value for  a state of system.*/
+    double energy();
+    /** Function which change the state of system (random pertubation).*/
+    void changedState();
+    /** Function to reverse to the previous state.*/
+    void reverseChangedState();
+    /** Simulated annealing procedure.  */
+    int run()
+    {
+        CV_Assert(initialT>finalT);
+        double Ti = initialT;
+        double previousEnergy = energy();
+        int exchange = 0;
+        while (Ti > finalT)
+        {
+            for (int i = 0; i < iterPerStep; i++)
+            {
+                changedState();
+                double newEnergy = energy();
+                if (newEnergy < previousEnergy)
+                {
+                    previousEnergy = newEnergy;
+                }
+                else
+                {
+                    double r = rEnergy.uniform(double(0.0), double(1.0));
+                    if (r < exp(-(newEnergy - previousEnergy) / Ti))
+                    {
+                        previousEnergy = newEnergy;
+                        exchange++;
+                    }
+                    else
+                        reverseChangedState();
+                }
+
+            }
+            Ti *= coolingRatio;
+        }
+        return exchange;
+    }
+    /** Set intial temperature of simulated annealing procedure.
+    *@param x new initial temperature. x\>0
+    */
+    void setInitialTemperature(double x)
+    {
+        CV_Assert(x>0);
+        initialT = x;
+    };
+    /** Set final temperature of simulated annealing procedure.
+    *@param x new final temperature value. 0\<x\<initial temperature
+    */
+    void setFinalTemperature(double x)
+    {
+        CV_Assert(x>0);
+        finalT = x;
+    };
+    /** Set setCoolingRatio of simulated annealing procedure : T(t) = coolingRatio * T(t-1).
+    * @param x new cooling ratio value. 0\<x\<1
+    */
+    void setCoolingRatio(double x)
+    {
+        CV_Assert(x>0 && x<1);
+        coolingRatio = x;
+    };
+    /** Set number iteration per temperature step.
+    * @param ite number of iteration per temperature step ite \> 0
+    */
+    void setIterPerStep(int ite)
+    {
+        CV_Assert(ite>0);
+        iterPerStep = ite;
+    }
+    struct Impl;
+protected :
+    void init()
+    {
+        initialT = 2;
+        finalT = 0.1;
+        coolingRatio = 0.95;
+        iterPerStep = 100;
+    }
+    RNG rEnergy;
+    double coolingRatio;
+    double initialT;
+    double finalT;
+    int iterPerStep;
```

Comment:
Implementation details should be moved out from public headers.
All details should be stored in `SimulatedAnnealingSolver::Impl`, which should be placed in .cpp file.